src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,668 @@
+        } else {
+            // z-index set to 2 to allow the SVG to render underneath the radio buttons
+            buttonDiv = d3.select('body').append('div').style('position', 'relative').style('z-index', 2)
+        }
+        if (id) {
+            buttonDiv.attr('id', id)
+        }
+
+        const primaryButtonDiv = buttonDiv.append('div')
+        primaryButtonDiv
+            .append('input')
+            .attr('type', 'radio')
+            .attr('id', primaryButtonID)
+            .attr('name', 'filter-buttons')
+            .attr('checked', true)
+        primaryButtonDiv
+            .append('label')
+            .attr('for', primaryButtonID)
+            .style('font-size', NodeConfig.primaryTextSize)
+            .style('font-weight', 'bold')
+            .text('Primary resources')
+
+        const allButtonDiv = buttonDiv.append('div')
+        allButtonDiv.append('input').attr('type', 'radio').attr('id', allButtonID).attr('name', 'filter-buttons')
+        allButtonDiv
+            .append('label')
+            .attr('for', allButtonID)
+            .style('font-size', NodeConfig.primaryTextSize)
+            .style('font-weight', 'bold')
+            .text('All resources')
+
+        return buttonDiv
+    }
+
+    // Render a GraphObject
+    private update(newGraphObject: GraphObject | undefined) {
+        // Never occurs. Possibly undefined args are used in calls within a listener declaration,
+        // but those listener events are only fired with verified GraphObjects
+        if (newGraphObject === undefined) {
+            return
+        }
+
+        // Clear the graph
+        this.g.selectAll('g').remove()
+
+        this.links = this.constructLinks(this.g, newGraphObject.links)
+        this.nodes = this.constructNodes(this.g, newGraphObject.nodes, this.iconPaths, this.vscodeApi)
+
+        // Implement drag behavior
+        this.nodes.call(
+            d3
+                .drag<SVGGElement, NodeDatum>()
+                .on('start', this.dragStarted)
+                .on('drag', this.dragged)
+                .on('end', this.dragEnded)
+        )
+        // Attach nodes & links to simulation, upon which specified forces are applied
+        this.simulation.nodes(newGraphObject.nodes).on('tick', () => this.ticked())
+        this.simulation.force<d3.ForceLink<NodeDatum, LinkDatum>>('link')?.links(newGraphObject.links)
+
+        // Give energy to a new render
+        this.simulation.alphaTarget(PhysicsConfig.reheatAlphaTarget).restart()
+    }
+
+    // Arrow function because this function is referenced, and it needs to be bound
+    private dragStarted = (event: d3.D3DragEvent<SVGGElement, NodeDatum, unknown>, d: NodeDatum): void => {
+        if (!event.active) {
+            this.simulation.alphaTarget(PhysicsConfig.reheatAlphaTarget).restart()
+        }
+        d.fx = d.x
+        d.fy = d.y
+    }
+
+    // Arrow function because this function is referenced, and it needs to be bound
```

Comment:
Probably don't need these comments. Not really telling me much.

Reply:
Just didn't want a future reader to ask "Why are these functions in particular arrow, but not the others?". But yeah I guess that's easy to figure out

Reply:
It's one of those things that looks weird at first if you're not used to JavaScript, but looks normal once you realize it's being used for binding. But comments should be reserved for code that looks weird even if you're pretty comfortable with the language.